File: algorithms/0006.zigzag-conversion/solution.rs
impl Solution {
    pub fn convert(s: String, num_rows: i32) -> String {
        let s_len = s.len();
        let num_rows = num_rows as usize;

        if num_rows == 1 || s_len <= num_rows {
            return s;
        }

        let mut rows = vec![vec![]; num_rows];
        let mut row = 0;
        let mut i = 1i32;
        for c in s.chars() {
            rows[row].push(c);

            if row == 0 {
                i = 1;
            } else if row == num_rows - 1 {
                i = -1;
            }

            row = (row as i32 + i) as usize;
        }

        return rows.iter().fold("".to_string(), |s, t| {
            format!("{}{}", s, t.iter().collect::<String>())
        });
    }
}

File: algorithms/0006.zigzag-conversion/solution.go
func convert(s string, numRows int) string {
	sLen := len(s)

	if numRows == 1 || sLen <= numRows {
		return s
	}

	rows := make([][]rune, numRows)
	row := 0
	i := 1

	for _, r := range s {
		rows[row] = append(rows[row], r)

		if row == 0 {
			i = 1
		} else if row == numRows-1 {
			i = -1
		}

		row = row + i
	}

	res := ""
	for _, runes := range rows {
		res += string(runes)
	}

	return res
}
